docs/madr/1-reassign-evolve.md

```diff
@@ -0,0 +1,166 @@
+# Reassigning ANTs as Evolve/Upgrade Mechanism
+
+- Status: Proposed
+- Approvers: [Ariel], [Dylan], [Atticus], [Phil], [David]
+- Date: [2025-01-30]
+- Authors: [Atticus]
+
+## Context and Problem Statement
+
+Currently to evolve the ANT we use AOS `Eval` - this introduces the problem of
+evaluating what the current version of the code is. If we maintain a variable in
+the state of the process containing a version number, that tells us only what
+that version number is. At any point the user - owner - of the ANT may `Eval`
+new code into the process resulting in changes to the process that may not
+reflect what we expect the version number we assigned to reflect. To overcome
+this we currently use dryrun interactions on APIs and analyze the results to see
+if the APIs perform as expected - this is very inefficient as dry run calls to
+CU's are expensive in networking client side, as well as for the CU.
+
+## Decision Drivers
+
+- Improve the performance of ANT loading client side.
+- Ensure the upgrade process is more robust by using immutable versioning, such
+  as module IDs.
+- Reduce computational load on CU infra.
+- Mitigate risks associated with reassigning ANT state and ARNS records.
+
+## Considered Options
+
+1. **Current Approach (Baseline)**
+   - Multiple dry runs per ANT.
+   - ETH compatibility checked via handler analysis.
+   - Upgrade involves loading Lua code.
+2. **Leverage Re-Assign Name and \_boot capabilites to upgrade ANTs**
+   - Use the module ID to evaluate process version and supported capabilites
+   - Upgrade via spawning a new ANT and reassigning ARNS records.
+
+## Decision Outcome
+
+### Chosen Option: **Leverage Re-Assign Name and \_boot capabilites to upgrade ANTs**
+
+With the ARIO process now supporting
+[`Reassign-Name`](https://github.com/ar-io/ar-io-ant-process/pull/26) and recent
+additions to add [`_boot`](https://github.com/ar-io/ar-io-ant-process/pull/57)
+to fork the process state to a whole new ANT, we can assign that new ANT as the
+registered ANT to the specified ArNS name in the registry, and using the module
+ID of the process to identify process version and capabilities, and solely the
+module ID.
+
+This optimized approach reduces computational overhead and improves caching for
+better performance. It also shifts to a more robust upgrade mechanism using
+immutable module IDs rather than version detection heuristics. We can do this
+now that we compile and publish our own WASM binaries, rather than using an AOS
+binary and loading our code into it.
```

Comment:
Additionally, we can store the latest module ids in a process - enabling support to push new ANT modules dynamically, as they are produced (e.g. publish new module, update the repository process, and arns app pulls latest modules from process to inform user of upgrades). 

TLDR - I'd like to understand more what we get from this approach

Reply:
[3a4a3c0](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/3a4a3c09e94e2f4827b2e5ac787629224ef80fd8)